calculate_average_thomaswue.sh

```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
+#
+#  Copyright 2023 The original authors
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+sdk use java 21.0.1-graal
+echo "Script executed from: ${PWD}"
+NATIVE_IMAGE_OPTS="--gc=epsilon"
+JAVA_OPTS=""
+
+if [ -z "${JVM_MODE}" ]; then
+    echo "Chosing native image mode, set JVM_MODE variable to select JVM mode."
+    if [ -f ./image_calculateaverage_thomaswue ]; then
+        echo 'Picking up existing native image.'
+    else
+        if [ -f ./profile_thomaswue.iprof ]; then
+            echo 'Picking up profiling information from thomaswue.iprof.'
+        else
+            echo 'Could not find profiling information, therefore it will be now regenerated.'
```

Comment:
Hey @thomaswue, thanks a lot for this, very interesting to see PGO!

In its current shape, I think though that it goes somewhat against the spirit of this challenge, specifically "Implementations must not rely on specifics of a given data set". Which I think we kinda do when leveraging profile data from the first run during following ones. Admittedly, the wording could be more precise, but the general spirit is that runs shouldn't benefit from outcomes from previous runs (as otherwise, when taken to the extreme, one could calculate results once store them in a file and then simply load the content of that file in the subsequent runs).

What would be acceptable to me though is creating the profile data at build time, not relying on the specific data set used during evaluation (similar to how creating a static CDS archive is in bounds). Would that be an option?

Reply:
Sure, absolutely. The profiling is only used to automate compiler tuning knowledge one would otherwise need to manually specify (e.g. "inline a lot in this method"). I can for example train on the existing test data files or check in a slightly larger test file with maybe 1k lines?

Another thing maybe in the spirit of the competition to generate interesting insights could be to run that same algorithm in 3 different ways - i.e. JIT, AOT, AOT+PGO.

Reply:
>  The profiling is only used to automate compiler tuning knowledge one would otherwise need to manually specify (e.g. "inline a lot in this method")

Yepp, that's what I kinda thought.

> check in a slightly larger test file with maybe 1k lines?

Yeah, if you could add a 1K file and the expected output (as obtained via _calculate_average_baseline.sh_) to _src/test/resources_ that would be perfect. It would extend the tests and you can use it for "training". Can the PGO data be generated via Maven? If so, that'd be best, but it's not a strict requirement. A one-off script to run before the evaluation would be ok, too.

> Another thing maybe in the spirit of the competition to generate interesting insights could be to run that same algorithm in 3 different ways - i.e. JIT, AOT, AOT+PGO.

Agreed. Thing is, I'm really overwhelmed by the number of submissions and I won't have any capacity for further increasing scope of this right now. But you're more than welcome to do this experiment and publish any interesting insights on your own end.

Reply:
OK, sounds good. As a quick fix, I have turned off PGO by default and instead added global tuning flags in a commit added to this PR. Maybe approaches that require running the application before should be kept separate in terms of evaluation anyway.

Agreed that running with different configs probably increases the scope too much. It could be interesting to do this after 31st of Jan for a few of the submissions.

Let me know if we can assist you with some of the work. Obviously a downside of the success of the project to have so much attention ;-).

Reply:
Ok, excellent, will evaluate this one tomorrow (need to install the native image tool first). In the mean time, could you make sure that running this one shows no differences to the expected output:

```
./test.sh thomaswue
```

We've added this one just earlier today so as to have at least some basic assurance of correctness for implementations. Thanks again!

Reply:
OK, will check. You do not need to install the native image tool, because it is included for the latest Graal versions. So just "sdk install java 21.0.1-graal" is enough.

Reply:
Had to make the script more quiet and fix a minor issue when accumulating results, but now "./test.sh thomaswue" succeeds (two new commits added). The script will generate on first invocation the image with some extra output though. Let me know if this is appropriate or there should be a different way to integrate this into the build process.

Reply:
I think to be honest that all that time cooking the image should be added to the overall result.

Reply:
Don't get me wrong. I think is great to learn new things and put Java to the test. Just an opinion. From what I'm seeing (not related to this) Java 21 is the new C! tailor made hashes and data structures, branchless programming, varhandles etc. I think it would be better a simple and idiomatic solution. Anyways, waiting for your solution and sure I'll take a look to learn this approach, have a nice weekend and keep it up!

Reply:
Yes, I think it would be interesting to have some categories such as most idiomatic solution or solution with smallest amount of statements. Then one can test different JVM implementations and how they fare for each. When asking only for the fastest, naturally one gets a solution that might look very close to C code. However, this at least shows imho that if necessary Java can also be used on such low level.

Reply:
Yeah, I can see that angle, OTOH, it's super-hard to impossible to objectively rate what's "most idiomatic". Probably something better suited for a blog post or a talk where one can explore these nuances. That said, I think there's somewhere an interesting threshold of how far comparatively "standard" Java solutions go up on the leaderboard (quite far in fact) before the more extreme solutions kick in. I.e. "idiomatic" gets you surprisingly far.

Reply:
> The script will generate on first invocation the image with some extra output though. The image file would have to be cleaned up when updating the code (e.g., after the evaluation). Let me know if this is appropriate or there should be a different way to integrate this into the build process.

@thomaswue, so IIUC, this still does the profiling with the actual dataset in the first run, right? Didn't you mean to provide a separate dataset file and use this one when creating the native image? I.e. the build (with profiling) should be clearly be separated from the evaluation run. Can we extract the part which builds the native image (including PGO) to a separate script and invoke this one to build the application (instead of the usual `mvn verify`)? 

So the flow would become this:

Build: 
```
ln -s src/test/data/yourprofilingdataset.txt measurements.txt
./your_native_build_script.sh # builds the native binary and PGO data
```

Evaluation (i.e. what I am doing):

```
rm -f measurements.txt
ln -s measurements_1B.txt measurements.txt

for i in {1..5}
do
    ./calculate_average_thomaswue.sh # just the actual launch command, no build steps
done
.
```

Reply:
The use of PGO is disabled (just had the option to enable via environment variable). To avoid confusion and make this simpler, I deleted the code for optionally using PGO for now. There is an "additional_build_step_thomaswue.sh" that builds the image. The "calculcate_average_thomaswue.sh" script runs in JVM mode if the file is not there and otherwise picks up the generated image. Let me know if this now works better for you.

Reply:
I have also updated the PR description to clarify the new behavior and that PGO is not in use for this version.

Reply:
Yepp, this LGTM now, thanks!